### PR TITLE
feat: Improve webpack-dev-server performance 

### PR DESCRIPTION
**Pull request by dmtrKovalenko**
## Before: 
image

## After:
image
 
So the timings, after all, are much better – this is for the initial request when we bundling the heaviest spec of @cypress/react (mui autocomplete).

image

### CODE REVIEW

File: npm/webpack-dev-server/src/makeWebpackConfig.ts [hunk 1]
```diff
@@ -1,28 +1,51 @@
 import { debug as debugFn } from 'debug'
 import * as path from 'path'
-import { Configuration } from 'webpack'
+import * as webpack from 'webpack'
 import { merge } from 'webpack-merge'
+import defaultWebpackConfig from './webpack.config'
+import LazyCompilePlugin from 'lazy-compile-webpack-plugin'
```
Comment (lmiller1990) [inline npm/webpack-dev-server/src/makeWebpackConfig.ts:6]: This is the one: https://github.com/liximomo/lazy-compile-webpack-plugin ?

I'm surprised more people are not using this 🤔 seems not well known

File: npm/react/cypress/plugins/index.js [hunk 2]
```diff
@@ -17,7 +19,7 @@ const webpackConfig = {
       {
         test: /\.(js|jsx|mjs|ts|tsx)$/,
         loader: 'babel-loader',
-        options: babelConfig,
+        options: { ...babelConfig, cacheDirectory: path.resolve(__dirname, '..', '..', '.babel-cache') },
```
Comment (lmiller1990) [inline npm/react/cypress/plugins/index.js:22]: What is this? Is this also to speed perf up?

File: npm/webpack-dev-server/src/makeWebpackConfig.ts [hunk 3]
```diff
@@ -1,28 +1,51 @@
...
+function getLazyCompilationWebpackConfig (options: MakeWebpackConfigOptions): webpack.Configuration {
+  if (options.disableLazyCompilation || !options.isOpenMode) {
+    return {}
+  }
+
+  switch (WEBPACK_MAJOR_VERSION) {
+    case 4:
+      return { plugins: [new LazyCompilePlugin()] }
+    case 5:
+      return { experiments: { lazyCompilation: true } } as webpack.Configuration
```
Comment (lmiller1990) [inline npm/webpack-dev-server/src/makeWebpackConfig.ts:38]: Right - so webpack 5 has lazy built in, but webpack 4 you need the plugin?

**Review by lmiller1990 — DISMISSED** [review]
The code seems fine, I did not get a chance to test it out locally - left some comments which are more general questions for my understanding.

File: npm/webpack-dev-server/src/makeWebpackConfig.ts (hunk 3 above)
Comment (dmtrKovalenko) [reply npm/webpack-dev-server/src/makeWebpackConfig.ts:38]: I think this should be from the code <3

File: npm/react/cypress/plugins/index.js (hunk 2 above)
Comment (elevatebart) [reply npm/react/cypress/plugins/index.js:22]: And show users how to optimize theirs.

**Review by elevatebart — APPROVED** [review]
I like it !!

**Comment by dmtrKovalenko** [comment]
There is also a small UI defect – because compilation happens on the flight (if it takes more than a second user seeing a blink to the white screen). 

https://user-images.githubusercontent.com/16926049/108690551-29b4d280-7503-11eb-94b9-7525c9ab01c9.mov

**Pull request merged**